apiserver/common/imagecommon/imagemetadata_test.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package imagecommon_test
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/testing"
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/apiserver/common/imagecommon"
+	"github.com/juju/juju/apiserver/params"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/state/cloudimagemetadata"
+	coretesting "github.com/juju/juju/testing"
+)
+
+type imageMetadataSuite struct {
+	st *mockState
+}
+
+var _ = gc.Suite(&imageMetadataSuite{})
+
+func (s *imageMetadataSuite) SetUpTest(c *gc.C) {
+	mCfg := testConfig(c)
+
+	s.st = &mockState{
+		Stub: testing.Stub{},
+		saveMetadata: func(m []cloudimagemetadata.Metadata) error {
+			return nil
+		},
+		modelCfg: func() (*config.Config, error) {
+			return mCfg, nil
+		},
+	}
+}
+
+func (s *imageMetadataSuite) TestSaveEmpty(c *gc.C) {
+	errs, err := imagecommon.Save(s.st, params.MetadataSaveParams{})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(errs, gc.HasLen, 0)
+	s.st.CheckCallNames(c, []string{}...)
+}
+
+func (s *imageMetadataSuite) TestSaveModelCfgFailed(c *gc.C) {
+	m := params.CloudImageMetadata{
+		Source: "custom",
+	}
+	msg := "save error"
+
+	s.st.modelCfg = func() (*config.Config, error) {
+		return nil, errors.New(msg)
+	}
+
+	errs, err := imagecommon.Save(s.st, params.MetadataSaveParams{
+		Metadata: []params.CloudImageMetadataList{{
+			Metadata: []params.CloudImageMetadata{m},
+		}},
+	})
+	c.Assert(errors.Cause(err), gc.ErrorMatches, msg)
+	c.Assert(errs, gc.IsNil)
+	s.st.CheckCallNames(c, "ModelConfig")
+}
+
+func (s *imageMetadataSuite) TestSave(c *gc.C) {
+	m := params.CloudImageMetadata{
+		Source: "custom",
+	}
+	msg := "save error"
+
+	saveCalls := 0
+	s.st.saveMetadata = func(m []cloudimagemetadata.Metadata) error {
```

Comment:
wouldn't this be more straight forward if you used the stub's error methods?

```go
s.st.SetErrors(nil, errors.New("save error"))
```

then later

```go
s.st.CheckCalls(c, []testing.Stub{
    {"ModelConfig", nil},
    {"SaveMetadata, []interface{expectedMetadata1}},
    {"SaveMetadata, []interface{expectedMetadata2}},
})
```